src/transformStateWithClones.js

```diff
@@ -7,7 +7,31 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const stateHistory = [];
+  let currentState = { ...state };
+
+  actions.forEach((action) => {
+    switch (action.type) {
+      case 'addProperties':
+        currentState = { ...currentState, ...action.extraData };
+        break;
+      case 'removeProperties':
+        action.keysToRemove.forEach((key) => {
+          if (currentState.hasOwnProperty(key)) {
+            delete currentState[key];
```

Comment:
According to the [CODE STYLE] checklist, you should avoid mutating objects. Although `delete` is used correctly here, it's better to create a new object without the keys to remove. This can be achieved using `Object.keys`, `filter`, and `reduce` to construct a new object without the specified keys.